firmware/port.c

```diff
@@ -211,6 +302,61 @@ void port_exec_async_complete(PortData* p, ExecStatus s) {
     port_step(p);
 }
 
+void usart_test_send(SercomId id, uint8_t* buf, uint32_t buf_len){
+    if (!(sercom(id)->USART.INTFLAG.reg & SERCOM_USART_INTFLAG_DRE)) {
+        invalid();
+        return;
+    }
+
+    while (sercom(id)->USART.SYNCBUSY.reg) {
+        // wait until not busy
+    }
+
+    for (uint32_t i = 0; i < buf_len; i++){
+        // write one byte
+        sercom(id)->USART.DATA.reg = buf[i];
+
+        while (!(sercom(id)->USART.INTFLAG.reg & SERCOM_USART_INTFLAG_TXC)) {
+            // wait until byte is written
+        }
+    }
+}
+
+void uart_send_data(PortData *p){
+    if (uart_buf.buf_len > 0) {
+        // copy data into port and send it
+        // buffer is full, send off data
+        // p->state = EXEC_DONE;
+
+        // pad 2 bytes at the beginning
+        // 1st byte indicates uart rx
+        // 2nd byte indicates uart rx number. 
+        // this also means rx number has to be <255
+        // p->reply_len += uart_buf.buf_len;
+        if (p->uart_data_pos == 255) {
+            // if this is the first time that uart data is 
+            // getting put into reply buf, pad by 2.
+            p->uart_data_pos = p->reply_len; // keep track of where we write uart data headers
```

Comment:
I don't understand the point of this `uart_data_pos` mechanism. Shouldn't it just compute the size upfront based on the amount of data in the ring buffer and the space left in `reply_buf`, and write the header bytes and the data all at once?